app/routes/list-project.tsx

```diff
@@ -0,0 +1,531 @@
+import {
+  MAX_AVATAR_SIZE_BYTES,
+  parseListProjectForm,
+} from "~/utils/project-submission";
+import { createNewPullRequest } from "~/github.server";
+import { getRepoOwnerAndName } from "~/utils/repo-url";
+
+export function links() {
+  return [
+    { rel: "stylesheet", href: formStyles },
+    { rel: "stylesheet", href: markdownStyles },
+  ];
+}
+
+/**
+ * Check that the user is logged in or redirect to the /login page
+ */
+export const loader: LoaderFunction = async ({ request }) => {
+  await getCurrentUserOrRedirect(request);
+  return json({});
+};
+
+/**
+ * This method is called when the form is submitted. It performs data validation
+ * and uploads files to GitHub before creating a PR. If all goes well, it
+ * redirects to a success page.
+ */
+export const action: ActionFunction = async ({ request }) => {
+  const currentUser = await getCurrentUser(request);
+  const accessToken = await getAccessToken(request);
+
+  if (!currentUser || !accessToken) {
+    throw new Error("You must be logged in to save your profile");
+  }
+
+  // TODO check whether we already have a project for this repository
+
+  // Parse the form
+  const { validationErrors, files, repoUrl } = await parseListProjectForm(
+    request,
+    currentUser
+  );
+
+  // If there are any validation errors, return early
+  if (validationErrors) {
+    return json({ validationErrors });
+  }
+
+  if (!repoUrl) {
+    throw new Error("Missing repo URL in new project form");
+  }
+
+  // Create a new pull request with the content of the form
+  const { owner: repoOwner, name: repoName } = getRepoOwnerAndName(repoUrl);
+  const { pullRequestUrl } = await createNewPullRequest(
+    accessToken,
+    files,
+    repoOwner,
+    repoName
+  );
+
+  return redirect(`/project-listed?pr=${encodeURIComponent(pullRequestUrl)}`);
+};
+
+type TagsState = {
+  languages: MultiValue<{ label: string; value: string }>;
+  currentlySeeking: MultiValue<{ label: string; value: string }>;
+  tags: MultiValue<{ label: string; value: string }>;
+};
+
+/**
+ * This (massive) form allows users to list their projects on Open Source Hub.
+ */
+const ListProject: FC = () => {
+  const transition = useTransition();
+  const actionData = useActionData();
+
+  // The 3 tag dropdowns are controlled components and we keep track of their state here
+  const [tags, setTags] = useState<TagsState>({
+    languages: [],
+    currentlySeeking: [],
+    tags: [],
+  });
+
+  const updateTags =
+    (key: keyof TagsState) =>
+    (updatedTags: MultiValue<{ label: string; value: string }>) => {
+      setTags({ ...tags, [key]: updatedTags });
+    };
+
+  const formRef = useRef<HTMLFormElement>(null);
+
+  // Store a Project that can be used to preview the form
+  const [projectPreview, setProjectPreview] = useState<Project>();
+
+  // Show/hide the preview modal
+  const [showPreview, setShowPreview] = useState(false);
+
+  // Display a preview of the new project
+  const displayPreview = () => {
```

Comment:
I added a way to preview the changes because it's a huge form and I don't want people to be surprised by anything.

The way this works is that we format some data into a `Project` object and pass it to a `ProjectPreview` component. That component renders a modal on top of the form, and inside that modal is the `ProjectTemplate`.